### PR TITLE
feat(integer): improve scalar shifts performances

### PR DESCRIPTION
**Pull request by tmontaigu**
This reworks the left/right scalar shift method.

This new implementation takes more advantage of the
radix representation property and use a combo
of shifting/moving blocks via a rotate_left/right
optionnaly followed by in-block shifting and propagation
to other blocks.

This new implementation requires the carries to be empty
(not sure what the preconditions were for the previous implementation)
and the output will also have clean carries.
Requiring empty carries allows to do the shift in a way
that scales well with the number of blocks as we can use truly parallel
operations.

This means that the time required to shift is dependent
in the shift value, which should not be a security problem
as its a clear value. (The previous implementation also
had timing that depended on the shift value)

There are two scenarios possible:
- The shift only require to move blocks -> its fast
- The shift requires moving + in-block shifting -> slower,
  but still faster than the previous implementation.

Worst case is when shift is less than the number of bits
in a block.

This also changes the type of the `shift` parameter from
`usize` to `u64` to be consistent with other scalar operations.

The follwing pseudo bench code on 64 bits
gives the following time ranges:

With changes:
```
unchecked_left: BenchStat {
    min: 16.743µs
    max: 526.518563ms
    mean: 150.77871ms
}
unchecked_left_parallelized: BenchStat {
    min: 17.291µs
    max: 94.408455ms
    mean: 30.092279ms
}
unchecked_right: BenchStat {
    min: 16.723µs
    max: 548.417332ms
    mean: 160.345234ms
}
unchecked_right_parallelized: BenchStat {
    min: 16.978µs
    max: 97.955322ms
    mean: 33.500562ms
}
Measured in 37.890296743s
```

Previous code:
```
unchecked_left: BenchStat {
    min: 1.055401595s
    max: 1.156574075s
    mean: 1.085648592s
}
unchecked_left_parallelized: BenchStat {
    min: 559.636545ms
    max: 630.83338ms
    mean: 584.747893ms
}
unchecked_right: BenchStat {
    min: 1.055041354s
    max: 2.314891255s
    mean: 1.644513996s
}
unchecked_right_parallelized: BenchStat {
    min: 562.017144ms
    max: 1.275945891s
    mean: 894.812286ms
}
Measured in 421.412913883s
```

```rust
use rand::Rng;
use std::time::Instant;
use tfhe::integer::{gen_keys, RadixClientKey};
use tfhe::shortint::parameters::PARAM_MESSAGE_2_CARRY_2;

const NB_CTXT: usize = 32;
const NB_TEST: usize = 100;

struct BenchStat {
    min: Option<std::time::Duration>,
    max: Option<std::time::Duration>,
    sum: std::time::Duration,
    count: u32,
}

impl BenchStat {
    fn update(&mut self, elapsed: std::time::Duration) {
        if self.min.is_none() {
            self.min = Some(elapsed);
        } else {
            self.min = self.min.map(|l| l.min(elapsed));
        }
        if self.max.is_none() {
            self.max = Some(elapsed);
        } else {
            self.max = self.max.map(|l| l.max(elapsed));
        }
        self.sum += elapsed;
        self.count += 1;
    }

    fn print(&self) {
        println!("BenchStat {{");
        println!("    min: {:?}", self.min.unwrap());
        println!("    max: {:?}", self.max.unwrap());
        println!("    mean: {:?}", self.sum / self.count);
        println!("}}");
    }
}

type ShiftType = u64;

fn main() {
    let mut unchecked_left_timing = BenchStat::default();
    let mut unchecked_left_parallelized_timing = BenchStat::default();
    let mut unchecked_right_timing = BenchStat::default();
    let mut unchecked_right_parallelized_timing = BenchStat::default();

    let total = Instant::now();

    let param = PARAM_MESSAGE_2_CARRY_2;
    let (cks, sks) = gen_keys(&param);
    let cks = RadixClientKey::from((cks, NB_CTXT));

    let mut rng = rand::thread_rng();

    //Nb of bits to shift
    let tmp_f64 = param.message_modulus.0 as f64;
    let nb_bits = tmp_f64.log2().floor() as usize * NB_CTXT;
    let modulus = (param.message_modulus.0 as u128).pow(NB_CTXT as u32);
    assert_eq!(nb_bits, 64);

    for i in 0..NB_TEST {
        println!("{} / {NB_TEST}", i + 1);
        let clear = rng.gen::<u128>() % modulus;
        let scalar = rng.gen::<u128>() % nb_bits as u128;

        println!("clear: {clear}, scalar: {scalar}");

        let ct = cks.encrypt(clear);

        {
            let before = Instant::now();
            let ct_res = sks.unchecked_scalar_left_shift(&ct, scalar as ShiftType);
            unchecked_left_timing.update(before.elapsed());
            //assert!(ct_res.block_carries_are_empty());
            let dec_res: u128 = cks.decrypt(&ct_res);
            assert_eq!((clear << scalar) % modulus, dec_res);

            let before = Instant::now();
            let ct_res = sks.unchecked_scalar_left_shift_parallelized(&ct, scalar as ShiftType);
            unchecked_left_parallelized_timing.update(before.elapsed());
            //assert!(ct_res.block_carries_are_empty());
            let dec_res: u128 = cks.decrypt(&ct_res);
            assert_eq!((clear << scalar) % modulus, dec_res);
        }

        {
            let before = Instant::now();
            let ct_res = sks.unchecked_scalar_right_shift(&ct, scalar as ShiftType);
            unchecked_right_timing.update(before.elapsed());
            // assert!(ct_res.block_carries_are_empty());
            let dec_res: u128 = cks.decrypt(&ct_res);
            assert_eq!((clear >> scalar) % modulus, dec_res);

            let before = Instant::now();
            let ct_res = sks.unchecked_scalar_right_shift_parallelized(&ct, scalar as ShiftType);
            unchecked_right_parallelized_timing.update(before.elapsed());
            //assert!(ct_res.block_carries_are_empty());
            let dec_res: u128 = cks.decrypt(&ct_res);
            assert_eq!((clear >> scalar) % modulus, dec_res);
        }
    }

    print!("unchecked_left: ");
    unchecked_left_timing.print();
    print!("unchecked_left_parallelized: ");
    unchecked_left_parallelized_timing.print();

    print!("unchecked_right: ");
    unchecked_right_timing.print();
    print!("unchecked_right_parallelized: ");
    unchecked_right_parallelized_timing.print();

    println!("Measured in {:?}", total.elapsed());
}
```

BREAKING CHANGE: parameter type changed from usize to u64

### CODE REVIEW

File: tfhe/src/integer/server_key/radix_parallel/shift.rs [hunk 1]
```diff
@@ -2,11 +2,18 @@ use crate::integer::ciphertext::RadixCiphertext;
...
 impl ServerKey {
     /// Computes homomorphically a right shift.
     ///
     /// The result is returned as a new ciphertext.
     ///
+    /// # Requirements
+    ///
+    /// - The blocks parameter's carry space have at least one more bit than message space
+    /// - The input ciphertext carry buffer is emtpy / clean
+    ///
```
Comment (tmontaigu) [inline tfhe/src/integer/server_key/radix_parallel/shift.rs:16]: Add the #output

File: tfhe/benches/integer/bench.rs [hunk 2]
```diff
@@ -731,6 +739,9 @@ criterion_group!(
     scalar_add_parallelized,
     scalar_sub_parallelized,
     scalar_mul_parallelized,
+    scalar_mul_parallelized,
```
Comment (IceTDrinker) [inline tfhe/benches/integer/bench.rs:742]: twice saclar_mul_parallelized

File: tfhe/src/integer/server_key/radix/mod.rs [hunk 3]
```diff
@@ -157,8 +157,18 @@ impl ServerKey {
...
+        &self,
+        ctxt: &mut RadixCiphertext<PBSOrder>,
+        start_index: usize,
+    ) {
         let len = ctxt.blocks.len();
-        for i in 0..len {
+        for i in start_index..len {
             self.propagate(ctxt, i);
         }
     }
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/mod.rs:174]: is partial propagate really dev only then ?

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 4]
```diff
@@ -117,50 +117,82 @@ impl ServerKey {
...
+        }
 
-        //if r == 0
-        if modified_remainder == nb_bits {
-            self.full_propagate(ct);
-            self.blockshift_right_assign(ct, quotient);
-        } else {
-            // B/2^u = (B*2^{p-r}) / (2^{p*(q+1)})
-            self.unchecked_scalar_left_shift_assign(ct, modified_remainder);
+        let num_bits_in_block = (self.key.message_modulus.0 as f64).log2() as u64;
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/shift.rs:137]: can self.key.message_modulus.0.ilog2() be enough ? it's the floor of the log2 though

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 5]
```diff
@@ -117,50 +117,82 @@ impl ServerKey {
...
+        let num_blocks = ct.blocks.len();
 
-            // We partially propagate in order to not lose information
-            self.partial_propagate(ct);
-            self.blockshift_right_assign(ct, 1_usize);
+        // rotate left as the blocks are from LSB to MSB
+        ct.blocks.rotate_left(rotations);
+        for block in &mut ct.blocks[num_blocks - rotations..] {
+            *block = self.key.create_trivial(0);
+        }
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/shift.rs:146]: I was wondering at some point, would be interesting to have a "fill_with_trivial(val: u64)" function on the blocks ?

File: tfhe/src/integer/server_key/radix_parallel/tests.rs [hunk 6]
```diff
@@ -834,25 +834,19 @@ fn integer_unchecked_scalar_left_shift(param: Parameters) {
     // message_modulus^vec_length
     let modulus = param.message_modulus.0.pow(NB_CTXT as u32) as u64;
 
-    //Nb of bits to shift
     let tmp_f64 = param.message_modulus.0 as f64;
     let nb_bits = tmp_f64.log2().floor() as usize * NB_CTXT;
 
     for _ in 0..NB_TEST {
         let clear = rng.gen::<u64>() % modulus;
+        let scalar = rng.gen::<u64>() % nb_bits as u64;
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests.rs:842]: don't we want to check the overflow behavior ? We can do checked shift on the clear values

same thing for all tests on shifts that have the same pattern

**Review by IceTDrinker — COMMENTED** [review]
Got a call, I'll finish the review of the core of the algorithm after it

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 5 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:146]: to avoid some allocations/assignments and re-use the existing memory

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 7]
```diff
@@ -117,50 +117,82 @@ impl ServerKey {
...
+                    .unchecked_apply_lookup_table_bivariate(current_block, previous_block, &lut)
+            })
+            .collect::<Vec<_>>();
+        self.key.unchecked_scalar_right_shift_assign(
+            &mut ct.blocks[num_blocks - rotations - 1],
+            shift_within_block as u8,
+        );
+        self.key
+            .message_extract_assign(&mut ct.blocks[num_blocks - rotations - 1]);
+        assert_eq!(partial_blocks.len(), ct.blocks.len() - rotations - 1);
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/shift.rs:188]: could we maybe create the mutable slice first and then compare its length to the partial_blocks len ? Would make the comparison a bit less confusing, or maybe use the izip! macro from core crypto I think it does some bound check (or zip checked) but I'm sure we have a zip that verifies iterators have the same len

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 8]
```diff
@@ -222,15 +263,76 @@ impl ServerKey {
...
     /// ```
     pub fn unchecked_scalar_left_shift_assign<PBSOrder: PBSOrderMarker>(
         &self,
         ct: &mut RadixCiphertext<PBSOrder>,
-        shift: usize,
+        shift: u64,
     ) {
-        let tmp = 1_u64 << shift;
-        self.smart_scalar_mul_assign(ct, tmp);
+        // see parallel implementation for a bit more details
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/shift.rs:274]: comments from right shift may apply here as well

File: tfhe/src/integer/server_key/radix_parallel/shift.rs [hunk 9]
```diff
@@ -68,35 +89,105 @@ impl ServerKey {
...
+        let (partial_blocks, block) =
+            rayon::join(create_blocks_using_bivariate_pbs, shift_last_block);
+
+        debug_assert_eq!(partial_blocks.len(), ct.blocks.len() - rotations - 1);
+        for (block, shifted_block) in ct.blocks[..num_blocks - rotations - 1]
+            .iter_mut()
+            .zip(partial_blocks.into_iter())
+        {
+            *block = shifted_block
         }
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/shift.rs:188]: could be parallelized ? not sure we win a lot

File: tfhe/src/integer/server_key/radix_parallel/shift.rs [hunk 10]
```diff
@@ -268,10 +365,99 @@ impl ServerKey {
...
+        // We now have to put the blocks back togeter
+        debug_assert_eq!(partial_blocks.len(), ct.blocks.len() - rotations - 1);
+        ct.blocks[rotations] = block;
+        for (block, shifted_block) in ct.blocks[rotations + 1..]
+            .iter_mut()
+            .zip(partial_blocks.into_iter())
+        {
+            *block = shifted_block
+        }
+
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/shift.rs:459]: could be parallelized potentially, not sure we would win much

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 5 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:146]: Yes that would be cool, that would have much impact on performances, but would make the code look nicer i think

File: tfhe/src/integer/server_key/radix/mod.rs (hunk 3 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/mod.rs:174]: I've left the comment as it was originally, but yes its not really a dev only (as in you use it to debug) function

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 11]
```diff
@@ -117,50 +117,82 @@ impl ServerKey {
...
+                    // bits when shifting right afterwards
+                    current_block <<= num_bits_in_block;
+                    previous_block <<= num_bits_in_block;
+
+                    current_block >>= shift_within_block;
+                    previous_block >>= shift_within_block;
+
+                    // The way of gettint carry / message is reversed compared
+                    // to the usual way but its normal
+                    let message_of_current_block = current_block >> num_bits_in_block;
```
[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:166]: no it wont overflow the full code is

```                    // left shift not to lose
                    // bits when shifting right afterwards
                    current_block <<= num_bits_in_block;
                    previous_block <<= num_bits_in_block;

                    current_block >>= shift_within_block;
                    previous_block >>= shift_within_block;
```

Where current_block and previous_block are in [0...message_modulus-1] 
and shift_within_block < num_bits_in_block, because shift_within_block  = shift % num_bits_in_block

The we first left shift by a value that is strictly greater than the value we are right shifting afterwards

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:166]: yeah but here you re shift by num_bits_in_block right ?

File: tfhe/src/integer/server_key/radix_parallel/shift.rs (hunk 10 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/shift.rs:459]: Entering a thread/switching context to a thread just to replace (replace + implicit drop) a value won't provide any

File: tfhe/src/integer/server_key/radix_parallel/shift.rs (hunk 9 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/shift.rs:188]: Entering a thread/switching context to a thread just to replace (replace + implicit drop) a value won't provide any gains

File: tfhe/src/integer/server_key/radix_parallel/shift.rs (hunk 10 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/shift.rs:459]: right 🙄

File: tfhe/src/integer/server_key/radix_parallel/tests.rs (hunk 6 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:842]: Yes we should have test for when the input shift is > nb_bits.

the current behavior is that implemented is (should be)  equivalent to `x.checked_shl(y).unwrap_or(0)`, that is, if the shift is bigger than nb_bits the bits are discarded. Which is what I'm accustomed to.

default for rust is a bit different, something like `3u8 << 232` won't compile in debug mode with 
```
 attempt to shift left by `232_i32`, which would overflow
```

In release mode it compiles and gets replaced with something like `3u8 << (232 & 7)` which as en quivialent
`3u8.wrapping_shl(232)`.

What default should be chose ?

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:842]: I would expect an overflowing shift to output 0 I think, I think that as long as it's well documented it should be fine

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 4 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:137]: It should work, but it would bump the rust requirement to 1.67

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:137]: ah damn, I think I put some ilog2 somewhere 🤦‍♂️

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:137]: we need to upgrade our rust version -_-

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:137]: So i can put it

File: tfhe/src/integer/server_key/radix/mod.rs (hunk 3 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/mod.rs]: changed the comment

**Comment by tmontaigu** [comment]
@slab-ci cpu_test

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 5 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:146]: actually we have a `create_trivial_assign` I should use that intead

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 11 above)
[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs]: I agree on the good result, I just wonder if we should prefer a shift which underflows or the division the same way you did for the other scalar shift :)

[the remark this answers is not in the document]
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs]: Visualizing shift works better in my brain than division, and the underflow here is defined behaviour

[the remark this answers is not in the document]
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs]: then all good

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 12]
```diff
@@ -222,15 +267,85 @@ impl ServerKey {
...
+            .collect::<Vec<_>>();
+
+        // We do this block separately as this one does not
+        // need to get incoming bits from it neighbour
+        self.key.unchecked_scalar_left_shift_assign(
+            &mut ct.blocks[rotations],
+            shift_within_block as u8,
+        );
+        // as left shift is not done via PBS we message extract
+        self.key.message_extract_assign(&mut ct.blocks[rotations]);
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/shift.rs:337]: let's perform a single PBS here maybe ?

File: tfhe/src/integer/server_key/radix_parallel/shift.rs [hunk 13]
```diff
@@ -268,10 +367,103 @@ impl ServerKey {
...
+        let (partial_blocks, block) =
+            rayon::join(create_blocks_using_bivariate_pbs, shift_last_block);
+
+        // We started with num_blocks, discarded 'rotations' blocks
+        // and did the last one separately
+        let expected_num_partial_block = ct.blocks.len() - rotations - 1;
+        assert_eq!(partial_blocks.len(), expected_num_partial_block);
+        ct.blocks[rotations] = block;
+        for (block, shifted_block) in ct.blocks[rotations + 1..]
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/shift.rs:459]: I still believe something like

let dest_blocks = &mut ct.blocks[rotations + 1..];
assert_eq!(partial_blocks.ln(), dest_blocks.len());
for ... in dest_blocks.iter_mut()...

or using izip!/zip checked would be more readable 🤔

File: tfhe/src/integer/server_key/radix_parallel/tests.rs [hunk 14]
```diff
@@ -866,98 +858,71 @@ fn integer_default_scalar_left_shift(param: Parameters) {
...
     // message_modulus^vec_length
     let modulus = param.message_modulus.0.pow(NB_CTXT as u32) as u64;
 
-    //Nb of bits to shift
-    let tmp_f64 = param.message_modulus.0 as f64;
-    let nb_bits = tmp_f64.log2().floor() as usize * NB_CTXT;
-
     for _ in 0..NB_TEST {
         let clear = rng.gen::<u64>() % modulus;
+        let scalar = rng.gen::<u32>();
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: sorry to be annoying but here most of the time we would be testing the oveflowing scalar shift, can we have two halves to this test ? one where we know we are not overflowing/underflowing and the other where we force the overflow/underflow ?

potentially being sure we test "in ciphertext shifts" by taking shift in 0..message_modulus would be good here

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: I think we have seen in the past that corner cases are an important part of our testing and here I don't think we are covering the different shift behaviors

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 12 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:337]: I mean perform shift + message extract in one go

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:337]: yeah we can

File: tfhe/src/integer/server_key/radix_parallel/tests.rs (hunk 14 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: should this be in to different test functions or in the same one ?

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: it could be the same test function as long as we have the 3/4 sub cases well defined

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: could be in their own blocks like
//Generic case
{

}

// Overlow case
{

}

// All small shifts
{
for shift in 0..message_modulus or carry_modulus {
}

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: testing for all small shift for each test will increase time required to test, is that ok ?

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:914]: hmm maybe we won't loop for those (like no need to repeat the test 10 times) ? I let you be the judge of that

File: tfhe/src/integer/server_key/radix/shift.rs [hunk 15]
```diff
@@ -117,50 +117,82 @@ impl ServerKey {
...
+            .key
+            .generate_accumulator(|x| (x % message_modulus) >> shift_within_block);
+        self.key
+            .apply_lookup_table_assign(&mut ct.blocks[num_blocks - rotations - 1], &lut);
+
+        // We started with num_blocks, discarded 'rotations' blocks
+        // and did the last one separately
+        let blocks_to_replace = &mut ct.blocks[..num_blocks - rotations - 1];
+        assert_eq!(partial_blocks.len(), blocks_to_replace.len());
+        for (block, shifted_block) in itertools::izip!(blocks_to_replace, partial_blocks) {
```
Comment (IceTDrinker) [inline tfhe/src/integer/server_key/radix/shift.rs:192]: sorry if I was unclear, izip was needed only if it was checking the interator lengths itself, I was fine with the way it was written before, I just found the length check hard to read

btw we have our own izip reimported in core to avoid the itertools dependency

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs:192]: ah so what should i do 😁?

Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:192]: if you like the izip! look then I woudl use the one hidden in core_crypto

otherwise the previous way was fine

Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix/shift.rs]: ok gonna import the izip from core_crypto

File: tfhe/src/integer/server_key/radix_parallel/tests.rs (hunk 14 above)
Comment (tmontaigu) [reply tfhe/src/integer/server_key/radix_parallel/tests.rs:988]: I put the small shift outside the main test loop

File: tfhe/src/integer/server_key/radix/shift.rs (hunk 12 above)
Comment (IceTDrinker) [reply tfhe/src/integer/server_key/radix/shift.rs:334]: looks like this one is not fused in a single PBS?

**Review by IceTDrinker — APPROVED** [review]
Pending the last missing PBS "merge" I think this looks good !

Thanks a lot 🙏
